### ISSUE TITLE
Projection with 'mip' method is broken for SPH datasets in yt-4.0

### ISSUE THREAD

**Issue by qobilidop** [yt-project/yt#2258]
### Bug report

**Bug summary**

@sheagk noticed that in yt-4.0 `ProjectionPlot` with `method='mip'` gives odd results.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
from scipy.stats import binned_statistic_2d
import yt

ds = yt.load('FIRE_M12i_ref11/snapshot_600.hdf5')

_, center = ds.find_max(('PartType0', 'density'))
width = ds.quan(300, 'kpc')
box = ds.box(center - width / 2, center + width / 2)

bins = (100, 100)
l_unit = 'kpc'
d_unit = 'Msun/pc**3'

proj = yt.ProjectionPlot(
    ds, 'z', ('gas', 'density'), method='mip',
    center=center, width=width, data_source=box, buff_size=bins,
)
proj.set_cmap('density', 'viridis')
proj.set_unit('density', d_unit)
proj.save('yt-proj-mip.png')
proj.show()

x = box['gas', 'x'].to_value(l_unit)
y = box['gas', 'y'].to_value(l_unit)
values = box['gas', 'density'].to_value(d_unit)
stats, _, _, _ = binned_statistic_2d(x, y, values, statistic='max', bins=bins)
plt.figure(figsize=(4, 3))
plt.imshow(stats, norm=LogNorm())
plt.colorbar()
plt.savefig('mpl.png', dpi=200)
plt.show()
```

**Actual outcome**

yt-proj-mip.png:
![image](https://user-images.githubusercontent.com/5176695/57350015-3b03d380-7111-11e9-97d4-b605b354b8d1.png)

mpl.png:
![image](https://user-images.githubusercontent.com/5176695/57350310-78b52c00-7112-11e9-8c21-343357dcfae6.png)

**Expected outcome**

The two plots shall look similar (probably under a transpose). But currently their colorbar ranges are drastically different.

**Version Information**
  * yt version: yt-4.0 at [c24cec7](https://github.com/yt-project/yt/commit/c24cec7a6531b7960cb54b8a6d0a407537bbb4f3)

**Comment by sheagk** [comment]
Thanks for posting this @qobilidop.  Just to further verify that there's something wrong here, here's what I get using yt3.5:

```
ds = yt.load('FIRE_M12i_ref11/snapshot_600.hdf5')
_, center = ds.find_max(('gas', 'density'))
proj = yt.ProjectionPlot(
    ds, 'z', ('gas', 'density'), method='mip',
    center=center, width=(300, 'kpc'))
proj.set_cmap('density', 'viridis')
proj.set_unit('density', 'Msun/pc**3')
```

![image](https://user-images.githubusercontent.com/5874877/57397378-68886580-7181-11e9-8beb-d76b7813218c.png)

**Comment by qobilidop** [comment]
@ngoldbaum (or someone else you know with the expertise) Could you have a look and point me to the relevant code? I get pretty lost seeking where the work is actually done.

**Comment by ngoldbaum** [comment]
@qobilidop SPH projections happen here: https://github.com/yt-project/yt/blob/b4acb2f2eed03d20e2ced7c1fb80e91047faab6e/yt/geometry/coordinates/cartesian_coordinates.py#L299-L357

You'll presumably need to write a `pixelize_sph_kernel_mip_projection` cython utility. I'm not exactly sure what an MIP projection means for SPH data so you'll probably also have to think about that.

**Comment by qobilidop** [comment]
@ngoldbaum Thanks for your reply. I know this is where SPH projection happens. But what confuses me is where the code lands on when `method='mip'` is specified. I could track down the code to this line:
https://github.com/yt-project/yt/blob/b4acb2f2eed03d20e2ced7c1fb80e91047faab6e/yt/data_objects/construction_data_containers.py#L191-L192

Do you know where `self.func` gets called eventually?

**Comment by ngoldbaum** [comment]
I don't think that code actually does anything. Instead it's a combination of how `dl` is calculated below that point:

https://github.com/yt-project/yt/blob/b4acb2f2eed03d20e2ced7c1fb80e91047faab6e/yt/data_objects/construction_data_containers.py#L507-L508

And the `method` keyword argument for the QuadTree cython class:

https://github.com/yt-project/yt/blob/b4acb2f2eed03d20e2ced7c1fb80e91047faab6e/yt/utilities/lib/quad_tree.pyx#L114-L115

For what you're working on the way the QuadTree projection works doesn't actually matter at all since SPH projections don't use that machinery. Instead you're going to need to write a thing that takes densities, smoothing lengths, and field values at the particle position and somehow produces an "MIP" projection. Like I said in my previous comment I'm not exactly sure what an MIP projection means for SPH data, so you'll also need to figure that out.

**Comment by qobilidop** [comment]
Thanks! I think that's what I'm looking for.

As to the semantic meaning of "MIP" in the context of SPH, @sheagk might have an idea. My naive thinking is to return the maximum field value of all the particles intersecting the pixel sightline at each pixel position.

If there isn't a good use case for a "MIP" projection in SPH viz. Probably the way to move forward is to disable the "MIP" method entirely to reduce user confusion.

**Comment by ngoldbaum** [comment]
That makes sense, except I wouldn't use the field values directly but instead the field values scaled by the SPH kernel at the closest point between the particle and the ray for that pixel, which should be the maximum field value "felt" by that ray.

One problem with this approach is that it ignores the contributions from multiple SPH particles, which might add up to a value bigger than the SPH-interpolated value from all the individual particles. I'm not sure how you'd handle that without discretizing the ray, calculating the SPH-interpolated field values at many locations along the ray, and then using the maximum value of those discretized estimates.

**Comment by sheagk** [comment]
Hey guys, I'm not sure I understand the exact problem here.  Maybe it's because I'm thinking specifically about density (or density-like) fields, but any field like that has a well-defined single value at every point in space, even in particle-based codes.  For SPH it's just the weighted sum of all the particles' value whose kernel intersects with that point; it's more complicated in MFM, moving mesh, etc., but I think the SPH-like solution would satisfy most use cases (and then it would also extrapolate well to non-gas particle species).  

Is the problem that we don't have a clear definition of where we should sample the field?  I guess maybe there needs to be an optional parameter of the number of points sampled along each ray, or it could default to the larger of the buff_size?

**Comment by ngoldbaum** [comment]
Right now there isn’t any ray-casting, everything is done in the image plane.

**Comment by ngoldbaum** [comment]
If you look in the block of code I linked to above in the cartesian geometry handler where SPH projections get handled, we call a cythong function called `pixelize_sph_kernel_projection`. This function smooths data directly onto a pixelized image using a projected 2D version of the smoothing kernel. For the MIP I guess we need to spatially order the data somehow and keep using the 3D kernel. There’s probably a way to find the maximum value for a field along a ray in the SPH formalism but I haven’t spent time to derive a formula...
